2953 activitylog sanitization

mfacar opened the pull request:
fixes #2953
fixes #2879
fixes #3094

This PR includes:

- [ ] Exclusion of  API endpoints that we don't want logging
- [ ] Refactor of ActivityLogParser and addition of decorators for the raw entries that we want to keep
- [ ] Migration to get rid of the already logged raw data
- [ ] Saving activitylog entries both in mongodb and filesystem as well under `log` folder.
- [ ] TTL index to expire activitylog entries a year after the date of creation
- [ ] Prevent creating UpdateLog for activitylog collection
- [ ] Searching for anonymous entries at Activity Log page

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

RafaPolit commented on app/api/activitylog/activitylogMiddleware.js:
```diff
@@ -1,19 +1,27 @@
+  '/api/recoverpassword',
+  '/api/documents/pdfInfo',
+  '/api/documents/download',
+  'api/attachments/download',
```
Please include the forward slash

txau commented on .gitignore:
```diff
@@ -29,6 +29,7 @@ app/api/csv/specs/*.jpg
 app/api/attachments/specs/uploads/toDelete.txt
 uploaded_documents/*
 custom_uploads/*
+activity_log/*
```
Why adding a new folder when we already have a `log` folder that is already configured in the paths for production? I don't think this is a bad idea, at the same time this forces our production code and self-installed instances to manage this new variable when we release this code.

mfacar replied:
Given the activity log files include in its name the tenant I'll write them into the log folder.

txau commented on app/api/activitylog/specs/activitylog.spec.js:
```diff
@@ -31,7 +31,7 @@ describe('activitylog', () => {
       expect(entries.length).toBe(5);
```
From the description of the text ` should return the entries, excluding certain routes`. I think we have just one special case `{ $ne: '/api/semantic-search/notify-updates' }`. IMO we should not log these entries nor filter them.

mfacar replied:
Sure, I didn't realize that. This test is not needed anymore given we are not logging this special case and we'll delete the old entries.

txau commented on app/api/attachments/routes.js:
```diff
@@ -96,6 +97,7 @@ export default app => {
     uploadMiddleware(attachmentsPath),
+    activitylogMiddleware,
     validation.validateRequest(
       Joi.object()
```
Maybe take the time to switch from Joi to AJV? In order to get rid of some tech debt with the boy-scout approach.

mfacar replied:
Although to move the validation to AJV is easy, to update the validation tests is not being straightforward, maybe it could be better to change this when we have a definition for #2934

txau commented on app/api/activitylog/helpers.js:
```diff
@@ -8,66 +12,158 @@ export const methods = {
+    this.description = entryValue.desc;
+    this.data = data;
+    this.action = entryValue.method ? entryValue.method : methods.update;
+    this.beautified = true;
```
Do we actually need this flag? As far as I'm understanding this flag is used for controlling rendering on client side. Can't we beautify all raws, even the raw ones so we don't have the condition and get cleaner code and a more straightforward approach?

mfacar replied:
I've removed the beautified flag adding de RAW mark at the parsing, now all entries have the same structure on the client-side.

txau commented on app/api/activitylog/specs/activitylog.spec.js:
```diff
@@ -47,53 +47,61 @@ describe('activitylog', () => {
+      it('should filter by method', async () => {
+        const { rows: entries } = await activitylog.get({ method: ['POST'] });
+        expect(entries.length).toBe(1);
+        expect(entries[0].method).toBe('POST');
```
Maybe simplify this code to a foreach so that we don't need to know how many there are, all we need to know is that all of them are POST. This way if the fixtures change the condition is better covered and still met.

txau commented on app/api/activitylog/specs/activitylog.spec.js:
```diff
@@ -47,53 +47,61 @@ describe('activitylog', () => {
-      expect(entries.length).toBe(1);
-      expect(entries[0].method).toBe('POST');
-    });
+    describe('filtering', () => {
```
Maybe switch the style to a more semantic approach, ie:
```
describe('when filtering')
  it('should return results by method')
```
So that it reads "when filtering it should return results by method".

mfacar commented on app/api/migrations/migrations/29-activity_log_sanitization/index.js:
```diff
@@ -0,0 +1,37 @@
+      `${deletedUploadEntriesWithoutBody.result.n} activity log POST entries deleted with empty bodies\r\n`
+    );
+
+    await db.collection('activitylogs').createIndex({ expireAt: 1 }, { expireAfterSeconds: 0 });
```
You're right, we only need to have into account that there is not a file that contains this older logs

txau replied:
You are right, I think it is ok to sacrifice that part of the logs. The activity log feature was first deployed on September 2019, so the users will be losing record for just a few days.

txau reviewed (approved):
LGTM

fnocetti commented on app/api/activitylog/activitylogParser.js:
```diff
@@ -1,245 +1,180 @@
-    'Resumed semantic search'
-  ),
-  MIGRATE: migrationLog,
+import * as helpers from 'api/activitylog/helpers';
```
@mfacar this file was basically rewritten, I think it's a good opportunity to make it TS. You think that's relatively quick to accomplish?

fnocetti commented on app/api/migrations/migrations/29-activity_log_sanitization/specs/29-activity_log_sanitization.spec.js:
```diff
@@ -0,0 +1,78 @@
+        .find({ url: { $in: BODY_REQUIRED_ENDPOINTS }, body: { entityId: 'entity1' } })
+        .count();
+
+      expect(ignoredEntries).toBe(0);
```
Shouldn't we have at least one result to ensure the migration is keeping those records?

mfacar replied:
Actually, this is the intention of this test, but there were a couple of errors, thanks I didn't notice it.

fnocetti commented on app/api/utils/specs/date.spec.js:
```diff
@@ -25,4 +25,11 @@ describe('date helper', () => {
   });
+
+  describe('add years to current date', () => {
+    const today = new Date();
```
I think hardcoded dates could be "more predictable" and prevent possible flakinesses? What do you think?

mfacar replied:
I'm using this utility method specifically to add to a year to the current date, I'm not sure how to hardcoded it in tests.
I could change the function passing a fixed date and then add it 1 year fixed, but I don't know if it could be useful in other contexts.

fnocetti commented on app/api/odm/model.ts:
```diff
@@ -138,7 +74,7 @@ export class OdmModel<T> {
 
 export function instanceModel<T = any>(collectionName: string, schema: mongoose.Schema) {
-  const logHelper = new UpdateLogHelper<T>(collectionName);
+  const logHelper = createUpdateLogHelper(collectionName);
```
This is missing the generic type T

mfacar replied:
Great, good catch!

fnocetti reviewed (changes requested):
@mfacar I still need to review the logHelper and its tests but I'm leaving some comments as for now. Please take a look! Thank you!

mfacar commented on app/api/migrations/migrations/29-activity_log_sanitization/specs/29-activity_log_sanitization.spec.js:
```diff
@@ -0,0 +1,78 @@
+        .find({ url: { $in: IGNORED_ENDPOINTS } })
+        .count();
+
+      expect(ignoredEntries).toBe(0);
```
Sure, I've added some additional fixtures and move into a different test the assert of remaining entries.

mfacar commented on app/api/activitylog/specs/activitylogParser.spec.js:
```diff
@@ -35,14 +35,17 @@ describe('Activitylog Parser', () => {
   }
 
   describe('getSemanticData', () => {
     it('should report as beautified: false if no translation present for the route', async () => {
```
done, I've changed the description with RAW

mfacar commented on app/api/activitylog/helpers.js:
```diff
@@ -8,66 +12,156 @@ export const methods = {
+  ...(builder.extra && { extra: builder.extra }),
+});
+
+export class ActivityLogBuilder {
```
done

mfacar commented on app/api/activitylog/activitylogParser.js:
```diff
@@ -1,245 +1,180 @@
-    'Resumed semantic search'
-  ),
-  MIGRATE: migrationLog,
+import * as helpers from 'api/activitylog/helpers';
```
I've changed to ts, only that I had to include several `any` in some places because the content of the body/query could be anything

fnocetti commented:
I finished the Code Review and it LGTM. I will proceed with the testing today!

fnocetti commented:
@mfacar I tested several cases and everything works well. I have just two comments:
- There is no migration for the updateLogs. For example, it is still showing the activityLogs updates.
- The activitylogs for public forms (anonymous) are not registering any body, is that on purpose?

If the previous points are desired, then this is ready to be merged in my opinion!

fnocetti commented:
@mfacar Sorry, I just noted one more thing: in the file where the logs are written there's no separator between entries, their written one after the other in a single line. I think we should have a separator in case we actually need to parse it.

mfacar commented:
> * There is no migration for the updateLogs. For example, it is still showing the activityLogs updates.
I've added a migration to delete these records.
> * The activitylogs for public forms (anonymous) are not registering any body, is that on purpose?
I moved the activitylogmiddleware to the /api/public endpoint and update the decorator in order to have more information, given I'm adding this endpoint to the required body ones, all the previous entries will be deleted in the migration, given these doesn't contain any information, it is part of the sanitization of old entries.

fnocetti reviewed (approved):
LGTM! Good job! Merging this now.
